runtime/src/iree/hal/utils/libmpi.c

```diff
@@ -30,42 +25,57 @@ static iree_status_t iree_hal_mpi_dynamic_symbols_resolve_all(
   return iree_ok_status();
 }
 
-// Load the dynamic library and the symbols
 iree_status_t iree_hal_mpi_library_load(
     iree_allocator_t host_allocator, iree_dynamic_library_t** out_library,
-    iree_hal_mpi_dynamic_symbols_t** out_syms) {
+    iree_hal_mpi_dynamic_symbols_t* out_syms) {
```

Comment:
is there any particular reason to have the two out_ arguments manage memory differently? `out_library` is a `**` and `out_syms` is now only a `*`.

Reply:
the library is a reference counted object while the symbols are a table in memory. The library is allocated and freed as an object while the symbols are populated in-place in the owner. This avoids an allocation for the symbols as whatever is going to get the library and the symbols already has storage for them. There's a few other ways we could slice this but until we got to the MPI channel implementation that'd be sharing this header I didn't want to mess with it too much. The intent is that one place retains the library and stores the symbols then a pointer to the symbols is passed around to all the various places that may use them (channels/etc). There will always be one owner as MPI needs Init/Finalize.